torch/csrc/autograd/engine.cpp

```diff
@@ -313,45 +319,30 @@ auto Engine::thread_main(
         thread_on_exception(local_graph_task, task.fn_, e);
       }
     }
-    // Notify downstream about the completion of tasks depending
-    // on both where the task was executed, and who owned the overall
-    // graph (in case of reentrant execution.)  See Note [Reentrant backwards].
-    auto base_owner = local_graph_task->owner_;
-    // Task from a non-worker thread. Easy case.
-    if (base_owner == NO_DEVICE) {
-      if (--local_graph_task->outstanding_tasks_ == 0) {
-        // Lock mutex to notify the GraphTask waiting on not_done_
-        std::lock_guard<std::mutex> lock(local_graph_task->mutex_);
-        local_graph_task->not_done_.notify_all();
-      }
-    } else {
-      // If it's a task initiated from this thread, decrease the counter, but
-      // don't do anything - loop condition will do all checks for us next.
-      if (base_owner == worker_device) {
-        --local_graph_task->outstanding_tasks_;
-        // Otherwise send a dummy function task to the owning thread just to
-        // ensure that it's not sleeping. If it has work, it might see that
-        // graph_task->outstanding_tasks_ == 0 before it gets to the task, but
-        // it's a no-op anyway.
-      } else if (base_owner != worker_device) {
-        if (--local_graph_task->outstanding_tasks_ == 0) {
-          // Synchronize outstanding_tasks_ with queue mutex
-          std::atomic_thread_fence(std::memory_order_release);
-          ready_queue_by_index(base_owner)
-              .push(NodeTask(local_graph_task, nullptr, InputBuffer(0)));
-        }
-      }
+
+    // Decrement the outstanding tasks.
+    --local_graph_task->outstanding_tasks_;
+
+    // Check if we've completed execution.
+    bool gt_completed = graph_task_completed(local_graph_task);
+    if (gt_completed) {
+      mark_graph_task_completed(local_graph_task);
     }
-  }
 
-  // When current_depth is 0 this worker thread is done and we need to notify
-  // the parent thread waiting on the graph_task
-  // NOTE: An edge case for this is when reentrant calls are repeatedly made in
-  // a thread which is at its maximum stack depth and they keep exiting right
-  // after. We will always switch to a new thread for each call, so, we'll keep
-  // oscillating between the two threads.
-  if (reentrant_thread && current_depth == 0) {
-    graph_task->not_done_.notify_all();
+    auto base_owner = local_graph_task->owner_;
```

Comment:
Why don't we need to notify the owner thread when it is not a worker anymore? Is it because it just returns a Future now and the `mark_graph_task_completed` above will notify on the future?
If it is the case, a comment would be useful.

Reply:
Yes, the future takes care of this now. I will add a comment.